openssl-dynamic/src/main/c/io_netty_tcnative_jni_ocsp_Native.c

```diff
@@ -0,0 +1,314 @@
+static jclass CALLBACK_CLAZZ = NULL;
+static jmethodID CALLBACK_METHOD = NULL;
+
+static void logException(JNIEnv *env) {
+  (*env)->ExceptionDescribe(env);
+  (*env)->ExceptionClear(env);
+}
+
+/**
+ * Frees the callback
+ */
+static void freeCallback(JNIEnv *env, tcn_ssl_ctxt_t *c, jweak callbackRef) {
+  
+  if (c != NULL) {
+    SSL_CTX_set_tlsext_status_arg(c->ctx, NULL);
+    SSL_CTX_set_tlsext_status_cb(c->ctx, NULL);
+  }
+
+  (*env)->DeleteWeakGlobalRef(env, callbackRef);
+}
+
+/**
+ * See:
+ *   io_netty_tcnative_jni_ocsp_client_callback
+ *   io_netty_tcnative_jni_ocsp_server_callback
+ */
+static int io_netty_tcnative_jni_ocsp_callback(int errorCode, SSL *ssl, void *arg) {
+
+  // This is conceptually not possible.
+  if (CALLBACK_CLAZZ == NULL) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no callback class, %i\n", errorCode);
+    return errorCode;
+  }
+
+  // This is conceptually not possible.
+  if (CALLBACK_METHOD == NULL) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no callback method, %i\n", errorCode);
+    return errorCode;
+  }
+
+  if (arg == NULL) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no callback object, %i\n", errorCode);
+    return errorCode;
+  }
+
+  JavaVM *jvm = tcn_get_java_vm();
+  if (jvm == NULL) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no JVM, %i\n", errorCode);
+    return errorCode;
+  }
+
+  JNIEnv *env;
+  if (tcn_get_java_env(&env) != JNI_OK) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no JNIEnv, %i\n", errorCode);
+    return errorCode;
+  }
+
+  if ((*jvm)->AttachCurrentThread(jvm, (void **)&env, NULL) != JNI_OK) {
+    fprintf(stderr, "AttachCurrentThread: %i\n", errorCode);
+    logException(env);
+    return errorCode;
+  }
+
+  jint value = 0;
+  jobject callback = (*env)->NewLocalRef(env, (jweak)arg);
+
+  if ((*env)->IsSameObject(env, callback, NULL) == JNI_FALSE) {
+
+    value = (*env)->CallIntMethod(env, callback, CALLBACK_METHOD, (jlong)ssl);
+
+    if ((*env)->ExceptionCheck(env)) {
+      fprintf(stderr, "CallIntMethod: %i\n", errorCode);
+      logException(env);
+
+      // Use the errorCode in the event of an JVM exception
+      value = errorCode;
+    }
+  }
+
+  (*env)->DeleteLocalRef(env, callback);
+  (*jvm)->DetachCurrentThread(jvm);
+
+  return (int)value;
+}
+
+/**
+ * The client uses -1 for errors
+ */
+static int io_netty_tcnative_jni_ocsp_client_callback(SSL *ssl, void *arg) {
+  return io_netty_tcnative_jni_ocsp_callback(-1, ssl, arg);
+}
+
+/**
+ * The server uses SSL_TLSEXT_ERR_ALERT_FATAL for errors (it's a non-negative number).
+ */
+static int io_netty_tcnative_jni_ocsp_server_callback(SSL *ssl, void *arg) {
+  return io_netty_tcnative_jni_ocsp_callback(SSL_TLSEXT_ERR_ALERT_FATAL, ssl, arg);
+}
+
+jint io_netty_tcnative_jni_ocsp_Native_OnLoad(JNIEnv *env) {
+  jclass clazz = (*env)->FindClass(env, "io/netty/tcnative/jni/ocsp/StatusCallback");
+  if ((*env)->ExceptionCheck(env)) {
```

Comment:
it should be good enough to just check `if (clazz == null) { return; }` right .. the exception should be pending and be thrown when we go back to java (we do this in the epoll module).